File: 5.Longest-Palindromic-Substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  let start = 0;
  let max_length = 1;

  const expandFromMiddle = (left, right) => {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      const current_length = right - left + 1;

      if (current_length > max_length) {
        max_length = current_length;
        start = left;
      }

      left -= 1;
      right += 1;
    }
  };

  for (let i = 0; i < s.length; i++) {
    expandFromMiddle(i - 1, i + 1); // odd length
    expandFromMiddle(i, i + 1); // even length
  }

  return s.substring(start, start + max_length);
};

File: python/5.Longest-Palindromic-Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        start = 0
        max_length = 1

        def expandFromMiddle(left, right):
            nonlocal start, max_length

            while left >= 0 and right < len(s) and s[left] == s[right]:
                cur_length = right - left + 1

                if cur_length > max_length:
                    max_length = cur_length
                    start = left

                left -= 1
                right += 1

        for i in range(len(s)):
            expandFromMiddle(i - 1, i + 1)
            expandFromMiddle(i, i + 1)

        return s[start: start + max_length]